### ISSUE TITLE
Don't typecheck final assignment as if it's literal

### ISSUE THREAD

**Issue by A5rocks** [python/mypy#15640]
**Bug Report**

Mypy typechecks assignment to finals. That's confusing, leads to more errors being printed, and loses out on (minuscule amounts of) performance

**To Reproduce**

```python
from typing import Final

x: Final = 2
x = "hi"
```

**Expected Behavior**

Output:
```
main.py:4: error: Cannot assign to final name "x"  [misc]
Found 1 error in 1 file (checked 1 source file)
```

**Actual Behavior**

Output:
```
main.py:4: error: Cannot assign to final name "x"  [misc]
main.py:4: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

https://mypy-play.net/?mypy=latest&python=3.11&gist=8af7cc283fa2811e2e86cdff159da3e9

**Comment by sobolevn** [comment]
Why do you think that the current state is not good? I see two problems:
- Reassignment
- Type change

Both are reported :)

**Comment by A5rocks** [comment]
Mostly cause it doesn't matter that the types are different. I dunno. It's more a matter of fact that technically:
```
x: Final = 2
x = 1 # this is a type error cause Literal[1] vs Literal[2]
```

And that's like, yeah that's true but if x weren't Final then it would be fine type wise too.

This came up cause one of my PRs makes this actually detect the type error xd

**Comment by sobolevn** [comment]
If you remove `Final` you get the similar results:

```python
x = 1
x = 'a'  # E: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
```

So, in my opinion, changing it might be really confusing.

**Comment by A5rocks** [comment]
I think that `last_known_value` should be trimmed off an instance if it's final and we're checking assignment. I think that is really what bothers me: more general different type errors are fine with me.

**Comment by erictraut** [comment]
Mypy's behavior here seems fine to me. The error messages are unambiguous and factually correct.
